### ISSUE TITLE
LocalOptGroup does not account for optimizers returning dicts

### ISSUE THREAD

**Issue by ajax2leet** [Theano/Theano#5296]
The transform function in LocalOptGroup fails, when an optimizer in the group returns a dict.
The issue is with this line:
https://github.com/Theano/Theano/blob/a7900bd8ddbfceda9249f235e21005e0ee970884/theano/gof/opt.py#L1324

**Comment by Sentient07** [comment]
Which optimizer returns a dict ? That part of the code says that with the list of Variables returned, if they don't have a parent, or is the input of the graph, we're not trying to optimize it any further. 
Can you give an instant where `new_repl` can be a dict ?

**Comment by ajax2leet** [comment]
e.g. local_fill_sink: https://github.com/Theano/Theano/blob/a7900bd8ddbfceda9249f235e21005e0ee970884/theano/tensor/opt.py#L1688

This is explicitely allowed as a return statement by the LocalOptimizer interface (https://github.com/Theano/Theano/blob/a7900bd8ddbfceda9249f235e21005e0ee970884/theano/gof/opt.py#L1061)

**Comment by Sentient07** [comment]
In the first link, it is not a `LocalOptGroup`. Its a `LocalOptimizer`. They're different. And in the second link, it is said that it could be one of them. Not a compulsory requirement.

**Comment by ajax2leet** [comment]
The transform function of LocalOptGroup loops over all optimizers, therefore `new_repl` can be an optimizer. And if this optimizer returns a dict (as it is allowed to do) the function will throw an exception (as there is no node "0")

**Comment by nouiz** [comment]
Ideally, we should support it. But in that case, I think we can only
support a subset when a dict is returned. If the node we try to replace is
the node the LocalOptGroup asked.

If it try to optimize something else, then I don't know how we could handle
that. We could try to call the following opt on that new node instead.

This is for your PR: https://github.com/Theano/Theano/pull/5272?

**Comment by Sentient07** [comment]
It doesn't loop over all the optimizers. For example, `gof.optdb.LocalGroupDB` registers into LocalOptGroup.

**Comment by ajax2leet** [comment]
Yes, this is for #5272. There the following LocalOptGroup is registered: ```gof.LocalOptGroup(local_log_sum_exp,
                                                local_mul_exp,
                                                # For clean up
                                                local_add_canonizer,
                                                local_fill_cut,
                                                local_fill_sink,
                                                local_pow_canonicalize,
                                                apply_all_opts=True)```

There LocalOptGroup.transform() fails on graphs like x**(2) * x**(-2)
